File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.py
class Solution(object):
    def longestPalindromeSubseq(self, s):
        """
        :type s: str
        :rtype: int
        """
        n=len(s)
        t=s[::-1]

        #dp=[[0]*(n+1) for _ in xrange(n+1)]

        prev=[0]*(n+1)

        for i in xrange(1,n+1):
            cur=[0]*(n+1)
            for j in xrange(1,n+1):
                if s[i-1]==t[j-1]:
                    cur[j]=1+prev[j-1]
                else:
                    cur[j]=max(prev[j],cur[j-1])
            prev=cur
        return prev[n]                    

File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.java
class Solution {
    public int longestPalindromeSubseq(String s) {
        // Reverse the string s
        String t = new StringBuilder(s).reverse().toString();
        int n = s.length();
        int[] prev = new int[n + 1];
        
        // Dynamic programming to find the longest palindromic subsequence
        for (int i = 1; i <= n; i++) {
            int[] cur = new int[n + 1];
            for (int j = 1; j <= n; j++) {
                if (s.charAt(i - 1) == t.charAt(j - 1)) {
                    cur[j] = 1 + prev[j - 1];
                } else {
                    cur[j] = Math.max(cur[j - 1], prev[j]);
                }
            }
            prev = cur;
        }
        
        return prev[n];
    }
}
